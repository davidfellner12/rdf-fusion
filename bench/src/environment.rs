use crate::BenchmarkingOptions;
use crate::benchmarks::BenchmarkName;
use crate::prepare::{PrepRequirement, prepare_run_closure, prepare_run_command};
use crate::prepare::{ensure_file_download, prepare_file_download};
use anyhow::bail;
use std::path::{Path, PathBuf};
use std::sync::Mutex;

/// Represents a context used to execute benchmarks.
pub struct RdfFusionBenchContext {
    /// General options for the benchmarks.
    options: BenchmarkingOptions,
    /// The path to the data dir.
    data_dir: Mutex<PathBuf>,
    /// The path to the current directory. This might be different from the `data_dir` if a
    /// benchmark is running.
    results_dir: Mutex<PathBuf>,
}

impl RdfFusionBenchContext {
    /// Creates a new [RdfFusionBenchContext].
    pub fn new(
        options: BenchmarkingOptions,
        data_dir: PathBuf,
        results_dir: PathBuf,
    ) -> Self {
        Self {
            options,
            data_dir: Mutex::new(data_dir),
            results_dir: Mutex::new(results_dir),
        }
    }

    /// Returns the [BenchmarkingOptions] for this context.
    pub fn options(&self) -> &BenchmarkingOptions {
        &self.options
    }

    /// Resolves a relative path `file` against the data directory.
    pub fn join_data_dir(&self, file: &Path) -> anyhow::Result<PathBuf> {
        if !file.is_relative() {
            bail!("Only relative paths can be resolved.")
        }

        Ok(self.data_dir.lock().unwrap().join(file))
    }

    /// Creates a new folder in the results directory and uses it until [Self::pop_dir] is
    /// called.
    ///
    /// This can be used to create folder hierarchies to separate the results of different
    /// benchmarks.
    #[allow(clippy::create_dir)]
    pub fn push_dir(&self, dir: &str) -> anyhow::Result<()> {
        let mut data_dir = self.data_dir.lock().unwrap();
        let mut results_dir = self.results_dir.lock().unwrap();

        data_dir.push(dir);
        results_dir.push(dir);

        Ok(())
    }

    /// Pops the last directory from the stack.
    pub fn pop_dir(&self) {
        let mut data_dir = self.data_dir.lock().unwrap();
        let mut results_dir = self.results_dir.lock().unwrap();

        data_dir.pop();
        results_dir.pop();
    }

    /// Creates a new bencher and modifies the context for this benchmark.
    pub fn create_benchmark_context(
        &self,
        benchmark_name: BenchmarkName,
    ) -> anyhow::Result<BenchmarkContext<'_>> {
        self.push_dir(&benchmark_name.dir_name())?;
        Ok(BenchmarkContext {
            context: self,
            benchmark_name,
        })
    }
}

/// A benchmarker that can be used to execute benchmarks.
///
/// It holds a reference to the current context to store its results.
pub struct BenchmarkContext<'ctx> {
    /// Reference to the benchmarking context.
    context: &'ctx RdfFusionBenchContext,
    /// Name of the benchmark that is being executed.
    benchmark_name: BenchmarkName,
}

impl<'ctx> BenchmarkContext<'ctx> {
    /// Returns a reference to the benchmarking context.
    pub fn parent(&self) -> &RdfFusionBenchContext {
        self.context
    }

    /// Returns the name of the benchmark that is being executed.
    pub fn benchmark_name(&self) -> BenchmarkName {
        self.benchmark_name
    }

    /// Returns the path to the results directory of this benchmark.
    pub fn data_dir(&self) -> PathBuf {
        self.context.data_dir.lock().unwrap().clone()
    }

    /// Returns the path to the results directory of this benchmark.
    pub fn results_dir(&self) -> PathBuf {
        self.context.results_dir.lock().unwrap().clone()
    }

    /// Prepares the context such that `requirement` is fulfilled.
    pub async fn prepare_requirement(
        &self,
        requirement: PrepRequirement,
    ) -> anyhow::Result<()> {
        match requirement {
            PrepRequirement::FileDownload {
                url,
                file_name,
                action,
            } => prepare_file_download(self, url, file_name, action).await,
            PrepRequirement::RunClosure { execute, .. } => {
                prepare_run_closure(self, &execute)
            }
            PrepRequirement::RunCommand {
                workdir,
                program,
                args,
                ..
            } => {
                let workdir = self.context.join_data_dir(&workdir)?;
                prepare_run_command(&workdir, &program, &args)
            }
        }
    }

    /// Ensures that the `requirement` is fulfilled in this context.
    pub fn ensure_requirement(&self, requirement: PrepRequirement) -> anyhow::Result<()> {
        match requirement {
            PrepRequirement::FileDownload { file_name, .. } => {
                ensure_file_download(self, file_name.as_path())
            }
            PrepRequirement::RunClosure {
                check_requirement, ..
            }
            | PrepRequirement::RunCommand {
                check_requirement, ..
            } => check_requirement(self),
        }
    }
}

/// Pops the results directory from the context when the bencher is dropped.
impl Drop for BenchmarkContext<'_> {
    fn drop(&mut self) {
        self.context.pop_dir();
    }
}
