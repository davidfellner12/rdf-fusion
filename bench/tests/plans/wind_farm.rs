use crate::plans::canonicalize_uuids;
use anyhow::Context;
use datafusion::physical_plan::displayable;
use datafusion::prelude::SessionConfig;
use insta::assert_snapshot;
use rdf_fusion::store::Store;
use rdf_fusion::{QueryExplanation, QueryOptions};
use rdf_fusion_bench::benchmarks::Benchmark;
use rdf_fusion_bench::benchmarks::windfarm::{
    NumTurbines, WindFarmBenchmark, WindFarmQueryName, get_wind_farm_raw_sparql_operation,
};
use rdf_fusion_bench::environment::{BenchmarkContext, RdfFusionBenchContext};
use rdf_fusion_bench::operation::SparqlRawOperation;
use std::path::PathBuf;

#[tokio::test]
pub async fn optimized_logical_plan_wind_farm() {
    for_all_explanations(|name, explanation| {
        assert_snapshot!(
            format!("{name} (Optimized)"),
            canonicalize_uuids(&explanation.optimized_logical_plan.to_string())
        )
    })
    .await;
}

#[tokio::test]
pub async fn execution_plan_wind_farm() {
    for_all_explanations(|name, explanation| {
        let string = displayable(explanation.execution_plan.as_ref())
            .indent(false)
            .to_string();
        assert_snapshot!(
            format!("{name} (Execution Plan)"),
            canonicalize_uuids(&string)
        )
    })
    .await;
}

async fn for_all_explanations(assertion: impl Fn(String, QueryExplanation) -> ()) {
    let benchmarking_context =
        RdfFusionBenchContext::new_for_criterion(PathBuf::from("./data"));

    // Load the benchmark data and set max query count to one.
    let benchmark = WindFarmBenchmark::new(NumTurbines::N4);
    let benchmark_name = benchmark.name();
    let benchmark_context = benchmarking_context
        .create_benchmark_context(benchmark_name)
        .unwrap();

    let store =
        Store::new_with_datafusion_config(SessionConfig::new().with_target_partitions(1));
    for query_name in WindFarmQueryName::list_queries() {
        let benchmark_name = format!("Wind Farm - {query_name}");
        let query = get_query_to_execute(&benchmark_context, query_name);

        let (_, explanation) = store
            .explain_query_opt(query.text(), QueryOptions::default())
            .await
            .unwrap();

        assertion(benchmark_name, explanation);
    }
}

fn get_query_to_execute(
    benchmark_context: &BenchmarkContext,
    query_name: WindFarmQueryName,
) -> SparqlRawOperation<WindFarmQueryName> {
    get_wind_farm_raw_sparql_operation(benchmark_context, query_name)
        .context("Could not list raw operations for Wind Farm benchmark. Have you prepared a wind-farm-4 dataset?")
        .unwrap()
}
