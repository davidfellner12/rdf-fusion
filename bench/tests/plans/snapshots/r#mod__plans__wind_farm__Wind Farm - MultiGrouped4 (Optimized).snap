---
source: bench/tests/plans/wind_farm.rs
expression: "&explanation.optimized_logical_plan.to_string()"
---
Projection: ENC_PT(site_label) AS site_label, ENC_PT(wtur_label) AS wtur_label, year AS year, month AS month, day AS day, hour AS hour, minute_10 AS minute_10, ENC_PT(val_prod_avg) AS val_prod_avg, ENC_PT(val_dir_avg) AS val_dir_avg, ENC_PT(val_speed_avg) AS val_speed_avg
  Projection: site_label, wtur_label, year, month, day, hour, minute_10, val_prod_avg, val_dir_avg, val_speed_avg
    Projection: site_label, wtur_label, year, month, day, hour, minute_10, <uuid>, <uuid>, <uuid>, val_prod_avg, val_dir_avg, <uuid> AS val_speed_avg
      Projection: site_label, wtur_label, year, month, day, hour, minute_10, <uuid>, <uuid>, <uuid>, val_prod_avg, <uuid> AS val_dir_avg
        Projection: site_label, wtur_label, year, month, day, hour, minute_10, <uuid>, <uuid>, <uuid>, <uuid> AS val_prod_avg
          Projection: site_label, wtur_label, year, month, day, hour, minute_10, <uuid>, <uuid>, <uuid>
            Aggregate: groupBy=[[site_label AS site_label, wtur_label AS wtur_label, ENC_PT(year) AS year, ENC_PT(month) AS month, ENC_PT(day) AS day, ENC_PT(hour) AS hour, ENC_PT(minute_10) AS minute_10]], aggr=[[AVG(ENC_TV(val_prod)) AS <uuid>, AVG(ENC_TV(val_dir)) AS <uuid>, AVG(ENC_TV(val_speed)) AS <uuid>]]
              Filter: EBV(BOOLEAN_AS_TERM(EBV(GEQ(ENC_TV(t), ENC_TV(Struct({term_type:2,value:2022-08-30T08:46:53,data_type:http://www.w3.org/2001/XMLSchema#dateTime,language_tag:})))) AND EBV(LEQ(ENC_TV(t), ENC_TV(Struct({term_type:2,value:2022-08-30T21:46:53,data_type:http://www.w3.org/2001/XMLSchema#dateTime,language_tag:}))))))
                Projection: site, site_label, wtur_asp, wtur_label, wtur, gensys_asp, gensys, generator_asp, generator, weather_asp, weather, ts_speed, dp_speed, val_speed, t, ts_dir, dp_dir, val_dir, ts_prod, dp_prod, val_prod, minute_10, hour, day, month, YEAR(ENC_TV(t)) AS year
                  Projection: site, site_label, wtur_asp, wtur_label, wtur, gensys_asp, gensys, generator_asp, generator, weather_asp, weather, ts_speed, dp_speed, val_speed, t, ts_dir, dp_dir, val_dir, ts_prod, dp_prod, val_prod, minute_10, hour, day, MONTH(ENC_TV(t)) AS month
                    Projection: site, site_label, wtur_asp, wtur_label, wtur, gensys_asp, gensys, generator_asp, generator, weather_asp, weather, ts_speed, dp_speed, val_speed, t, ts_dir, dp_dir, val_dir, ts_prod, dp_prod, val_prod, minute_10, hour, DAY(ENC_TV(t)) AS day
                      Projection: site, site_label, wtur_asp, wtur_label, wtur, gensys_asp, gensys, generator_asp, generator, weather_asp, weather, ts_speed, dp_speed, val_speed, t, ts_dir, dp_dir, val_dir, ts_prod, dp_prod, val_prod, minute_10, HOURS(ENC_TV(t)) AS hour
                        Projection: site, site_label, wtur_asp, wtur_label, wtur, gensys_asp, gensys, generator_asp, generator, weather_asp, weather, ts_speed, dp_speed, val_speed, t, ts_dir, dp_dir, val_dir, ts_prod, dp_prod, val_prod, MUL(ENC_TV(Struct({term_type:2,value:10,data_type:http://www.w3.org/2001/XMLSchema#integer,language_tag:})), FLOOR(DIV(MINUTES(ENC_TV(t)), ENC_TV(Struct({term_type:2,value:10.0,data_type:http://www.w3.org/2001/XMLSchema#decimal,language_tag:}))))) AS minute_10
                          Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys, lhs.generator_asp AS generator_asp, lhs.generator AS generator, lhs.weather_asp AS weather_asp, lhs.weather AS weather, lhs.ts_speed AS ts_speed, lhs.dp_speed AS dp_speed, lhs.val_speed AS val_speed, lhs.t AS t, lhs.ts_dir AS ts_dir, lhs.dp_dir AS dp_dir, lhs.val_dir AS val_dir, lhs.ts_prod AS ts_prod, lhs.dp_prod AS dp_prod, lhs.val_prod AS val_prod
                            Inner Join: lhs.dp_prod = rhs.dp_prod, lhs.t = rhs.t
                              SubqueryAlias: lhs
                                Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys, lhs.generator_asp AS generator_asp, lhs.generator AS generator, lhs.weather_asp AS weather_asp, lhs.weather AS weather, lhs.ts_speed AS ts_speed, lhs.dp_speed AS dp_speed, lhs.val_speed AS val_speed, lhs.t AS t, lhs.ts_dir AS ts_dir, lhs.dp_dir AS dp_dir, lhs.val_dir AS val_dir, lhs.ts_prod AS ts_prod, lhs.dp_prod AS dp_prod, rhs.val_prod AS val_prod
                                  Inner Join: lhs.dp_prod = rhs.dp_prod
                                    SubqueryAlias: lhs
                                      Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys, lhs.generator_asp AS generator_asp, lhs.generator AS generator, lhs.weather_asp AS weather_asp, lhs.weather AS weather, lhs.ts_speed AS ts_speed, lhs.dp_speed AS dp_speed, lhs.val_speed AS val_speed, lhs.t AS t, lhs.ts_dir AS ts_dir, lhs.dp_dir AS dp_dir, lhs.val_dir AS val_dir, lhs.ts_prod AS ts_prod, rhs.dp_prod AS dp_prod
                                        Inner Join: lhs.ts_prod = rhs.ts_prod
                                          SubqueryAlias: lhs
                                            Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys, lhs.generator_asp AS generator_asp, lhs.generator AS generator, lhs.weather_asp AS weather_asp, lhs.weather AS weather, lhs.ts_speed AS ts_speed, lhs.dp_speed AS dp_speed, lhs.val_speed AS val_speed, lhs.t AS t, lhs.ts_dir AS ts_dir, lhs.dp_dir AS dp_dir, lhs.val_dir AS val_dir, lhs.ts_prod AS ts_prod
                                              Inner Join: lhs.ts_prod = rhs.ts_prod
                                                SubqueryAlias: lhs
                                                  Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys, lhs.generator_asp AS generator_asp, lhs.generator AS generator, lhs.weather_asp AS weather_asp, lhs.weather AS weather, lhs.ts_speed AS ts_speed, lhs.dp_speed AS dp_speed, lhs.val_speed AS val_speed, lhs.t AS t, lhs.ts_dir AS ts_dir, lhs.dp_dir AS dp_dir, lhs.val_dir AS val_dir, rhs.ts_prod AS ts_prod
                                                    Inner Join: lhs.generator = rhs.generator
                                                      SubqueryAlias: lhs
                                                        Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys, lhs.generator_asp AS generator_asp, lhs.generator AS generator, lhs.weather_asp AS weather_asp, lhs.weather AS weather, lhs.ts_speed AS ts_speed, lhs.dp_speed AS dp_speed, lhs.val_speed AS val_speed, lhs.t AS t, lhs.ts_dir AS ts_dir, lhs.dp_dir AS dp_dir, lhs.val_dir AS val_dir
                                                          Inner Join: lhs.ts_dir = rhs.ts_dir
                                                            SubqueryAlias: lhs
                                                              Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys, lhs.generator_asp AS generator_asp, lhs.generator AS generator, lhs.weather_asp AS weather_asp, lhs.weather AS weather, lhs.ts_speed AS ts_speed, lhs.dp_speed AS dp_speed, lhs.val_speed AS val_speed, lhs.t AS t, lhs.ts_dir AS ts_dir, lhs.dp_dir AS dp_dir, lhs.val_dir AS val_dir
                                                                Inner Join: lhs.dp_dir = rhs.dp_dir, lhs.t = rhs.t
                                                                  SubqueryAlias: lhs
                                                                    Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys, lhs.generator_asp AS generator_asp, lhs.generator AS generator, lhs.weather_asp AS weather_asp, lhs.weather AS weather, lhs.ts_speed AS ts_speed, lhs.dp_speed AS dp_speed, lhs.val_speed AS val_speed, lhs.t AS t, lhs.ts_dir AS ts_dir, lhs.dp_dir AS dp_dir, rhs.val_dir AS val_dir
                                                                      Inner Join: lhs.dp_dir = rhs.dp_dir
                                                                        SubqueryAlias: lhs
                                                                          Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys, lhs.generator_asp AS generator_asp, lhs.generator AS generator, lhs.weather_asp AS weather_asp, lhs.weather AS weather, lhs.ts_speed AS ts_speed, lhs.dp_speed AS dp_speed, lhs.val_speed AS val_speed, lhs.t AS t, lhs.ts_dir AS ts_dir, rhs.dp_dir AS dp_dir
                                                                            Inner Join: lhs.ts_dir = rhs.ts_dir
                                                                              SubqueryAlias: lhs
                                                                                Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys, lhs.generator_asp AS generator_asp, lhs.generator AS generator, lhs.weather_asp AS weather_asp, lhs.weather AS weather, lhs.ts_speed AS ts_speed, lhs.dp_speed AS dp_speed, lhs.val_speed AS val_speed, lhs.t AS t, rhs.ts_dir AS ts_dir
                                                                                  Inner Join: lhs.weather = rhs.weather
                                                                                    SubqueryAlias: lhs
                                                                                      Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys, lhs.generator_asp AS generator_asp, lhs.generator AS generator, lhs.weather_asp AS weather_asp, lhs.weather AS weather, lhs.ts_speed AS ts_speed, lhs.dp_speed AS dp_speed, lhs.val_speed AS val_speed, lhs.t AS t
                                                                                        Inner Join: lhs.ts_speed = rhs.ts_speed
                                                                                          SubqueryAlias: lhs
                                                                                            Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys, lhs.generator_asp AS generator_asp, lhs.generator AS generator, lhs.weather_asp AS weather_asp, lhs.weather AS weather, lhs.ts_speed AS ts_speed, lhs.dp_speed AS dp_speed, lhs.val_speed AS val_speed, rhs.t AS t
                                                                                              Inner Join: lhs.dp_speed = rhs.dp_speed
                                                                                                SubqueryAlias: lhs
                                                                                                  Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys, lhs.generator_asp AS generator_asp, lhs.generator AS generator, lhs.weather_asp AS weather_asp, lhs.weather AS weather, lhs.ts_speed AS ts_speed, lhs.dp_speed AS dp_speed, rhs.val_speed AS val_speed
                                                                                                    Inner Join: lhs.dp_speed = rhs.dp_speed
                                                                                                      SubqueryAlias: lhs
                                                                                                        Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys, lhs.generator_asp AS generator_asp, lhs.generator AS generator, lhs.weather_asp AS weather_asp, lhs.weather AS weather, lhs.ts_speed AS ts_speed, rhs.dp_speed AS dp_speed
                                                                                                          Inner Join: lhs.ts_speed = rhs.ts_speed
                                                                                                            SubqueryAlias: lhs
                                                                                                              Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys, lhs.generator_asp AS generator_asp, lhs.generator AS generator, lhs.weather_asp AS weather_asp, lhs.weather AS weather, rhs.ts_speed AS ts_speed
                                                                                                                Inner Join: lhs.weather = rhs.weather
                                                                                                                  SubqueryAlias: lhs
                                                                                                                    Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys, lhs.generator_asp AS generator_asp, lhs.generator AS generator, lhs.weather_asp AS weather_asp, lhs.weather AS weather
                                                                                                                      Inner Join: lhs.weather = rhs.weather
                                                                                                                        SubqueryAlias: lhs
                                                                                                                          Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys, lhs.generator_asp AS generator_asp, lhs.generator AS generator, lhs.weather_asp AS weather_asp, rhs.weather AS weather
                                                                                                                            Inner Join: lhs.weather_asp = rhs.weather_asp
                                                                                                                              SubqueryAlias: lhs
                                                                                                                                Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys, lhs.generator_asp AS generator_asp, lhs.generator AS generator, rhs.weather_asp AS weather_asp
                                                                                                                                  Inner Join: lhs.wtur = rhs.wtur
                                                                                                                                    SubqueryAlias: lhs
                                                                                                                                      Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys, lhs.generator_asp AS generator_asp, lhs.generator AS generator
                                                                                                                                        Inner Join: lhs.generator = rhs.generator
                                                                                                                                          SubqueryAlias: lhs
                                                                                                                                            Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys, lhs.generator_asp AS generator_asp, rhs.generator AS generator
                                                                                                                                              Inner Join: lhs.generator_asp = rhs.generator_asp
                                                                                                                                                SubqueryAlias: lhs
                                                                                                                                                  Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys, rhs.generator_asp AS generator_asp
                                                                                                                                                    Inner Join: lhs.gensys = rhs.gensys
                                                                                                                                                      SubqueryAlias: lhs
                                                                                                                                                        Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, lhs.gensys AS gensys
                                                                                                                                                          Inner Join: lhs.gensys = rhs.gensys
                                                                                                                                                            SubqueryAlias: lhs
                                                                                                                                                              Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, lhs.gensys_asp AS gensys_asp, rhs.gensys AS gensys
                                                                                                                                                                Inner Join: lhs.gensys_asp = rhs.gensys_asp
                                                                                                                                                                  SubqueryAlias: lhs
                                                                                                                                                                    Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur, rhs.gensys_asp AS gensys_asp
                                                                                                                                                                      Inner Join: lhs.wtur = rhs.wtur
                                                                                                                                                                        SubqueryAlias: lhs
                                                                                                                                                                          Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, lhs.wtur AS wtur
                                                                                                                                                                            Inner Join: lhs.wtur = rhs.wtur
                                                                                                                                                                              SubqueryAlias: lhs
                                                                                                                                                                                Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, lhs.wtur_label AS wtur_label, rhs.wtur AS wtur
                                                                                                                                                                                  Inner Join: lhs.wtur_asp = rhs.wtur_asp
                                                                                                                                                                                    SubqueryAlias: lhs
                                                                                                                                                                                      Projection: lhs.site AS site, lhs.site_label AS site_label, lhs.wtur_asp AS wtur_asp, rhs.wtur_label AS wtur_label
                                                                                                                                                                                        Inner Join: lhs.wtur_asp = rhs.wtur_asp
                                                                                                                                                                                          SubqueryAlias: lhs
                                                                                                                                                                                            Projection: lhs.site AS site, lhs.site_label AS site_label, rhs.wtur_asp AS wtur_asp
                                                                                                                                                                                              Inner Join: lhs.site = rhs.site
                                                                                                                                                                                                SubqueryAlias: lhs
                                                                                                                                                                                                  Projection: lhs.site AS site, rhs.site_label AS site_label
                                                                                                                                                                                                    Inner Join: lhs.site = rhs.site
                                                                                                                                                                                                      SubqueryAlias: lhs
                                                                                                                                                                                                        QuadPattern (?site <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://github.com/magbak/chrontext/rds_power#Site>)
                                                                                                                                                                                                      SubqueryAlias: rhs
                                                                                                                                                                                                        QuadPattern (?site <http://www.w3.org/2000/01/rdf-schema#label> ?site_label)
                                                                                                                                                                                                SubqueryAlias: rhs
                                                                                                                                                                                                  QuadPattern (?site <https://github.com/magbak/chrontext/rds_power#hasFunctionalAspect> ?wtur_asp)
                                                                                                                                                                                          SubqueryAlias: rhs
                                                                                                                                                                                            QuadPattern (?wtur_asp <http://www.w3.org/2000/01/rdf-schema#label> ?wtur_label)
                                                                                                                                                                                    SubqueryAlias: rhs
                                                                                                                                                                                      QuadPattern (?wtur <https://github.com/magbak/chrontext/rds_power#hasFunctionalAspectNode> ?wtur_asp)
                                                                                                                                                                              SubqueryAlias: rhs
                                                                                                                                                                                QuadPattern (?wtur <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://github.com/magbak/chrontext/rds_power#A>)
                                                                                                                                                                        SubqueryAlias: rhs
                                                                                                                                                                          QuadPattern (?wtur <https://github.com/magbak/chrontext/rds_power#hasFunctionalAspect> ?gensys_asp)
                                                                                                                                                                  SubqueryAlias: rhs
                                                                                                                                                                    QuadPattern (?gensys <https://github.com/magbak/chrontext/rds_power#hasFunctionalAspectNode> ?gensys_asp)
                                                                                                                                                            SubqueryAlias: rhs
                                                                                                                                                              QuadPattern (?gensys <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://github.com/magbak/chrontext/rds_power#RA>)
                                                                                                                                                      SubqueryAlias: rhs
                                                                                                                                                        QuadPattern (?gensys <https://github.com/magbak/chrontext/rds_power#hasFunctionalAspect> ?generator_asp)
                                                                                                                                                SubqueryAlias: rhs
                                                                                                                                                  QuadPattern (?generator <https://github.com/magbak/chrontext/rds_power#hasFunctionalAspectNode> ?generator_asp)
                                                                                                                                          SubqueryAlias: rhs
                                                                                                                                            QuadPattern (?generator <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://github.com/magbak/chrontext/rds_power#GAA>)
                                                                                                                                    SubqueryAlias: rhs
                                                                                                                                      QuadPattern (?wtur <https://github.com/magbak/chrontext/rds_power#hasFunctionalAspect> ?weather_asp)
                                                                                                                              SubqueryAlias: rhs
                                                                                                                                QuadPattern (?weather <https://github.com/magbak/chrontext/rds_power#hasFunctionalAspectNode> ?weather_asp)
                                                                                                                        SubqueryAlias: rhs
                                                                                                                          QuadPattern (?weather <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://github.com/magbak/chrontext/rds_power#LE>)
                                                                                                                  SubqueryAlias: rhs
                                                                                                                    QuadPattern (?weather <https://github.com/magbak/chrontext#hasTimeseries> ?ts_speed)
                                                                                                            SubqueryAlias: rhs
                                                                                                              QuadPattern (?ts_speed <https://github.com/magbak/chrontext#hasDataPoint> ?dp_speed)
                                                                                                      SubqueryAlias: rhs
                                                                                                        QuadPattern (?dp_speed <https://github.com/magbak/chrontext#hasValue> ?val_speed)
                                                                                                SubqueryAlias: rhs
                                                                                                  QuadPattern (?dp_speed <https://github.com/magbak/chrontext#hasTimestamp> ?t)
                                                                                          SubqueryAlias: rhs
                                                                                            QuadPattern (?ts_speed <http://www.w3.org/2000/01/rdf-schema#label> "Windspeed")
                                                                                    SubqueryAlias: rhs
                                                                                      QuadPattern (?weather <https://github.com/magbak/chrontext#hasTimeseries> ?ts_dir)
                                                                              SubqueryAlias: rhs
                                                                                QuadPattern (?ts_dir <https://github.com/magbak/chrontext#hasDataPoint> ?dp_dir)
                                                                        SubqueryAlias: rhs
                                                                          QuadPattern (?dp_dir <https://github.com/magbak/chrontext#hasValue> ?val_dir)
                                                                  SubqueryAlias: rhs
                                                                    QuadPattern (?dp_dir <https://github.com/magbak/chrontext#hasTimestamp> ?t)
                                                            SubqueryAlias: rhs
                                                              QuadPattern (?ts_dir <http://www.w3.org/2000/01/rdf-schema#label> "WindDirection")
                                                      SubqueryAlias: rhs
                                                        QuadPattern (?generator <https://github.com/magbak/chrontext#hasTimeseries> ?ts_prod)
                                                SubqueryAlias: rhs
                                                  QuadPattern (?ts_prod <http://www.w3.org/2000/01/rdf-schema#label> "Production")
                                          SubqueryAlias: rhs
                                            QuadPattern (?ts_prod <https://github.com/magbak/chrontext#hasDataPoint> ?dp_prod)
                                    SubqueryAlias: rhs
                                      QuadPattern (?dp_prod <https://github.com/magbak/chrontext#hasValue> ?val_prod)
                              SubqueryAlias: rhs
                                QuadPattern (?dp_prod <https://github.com/magbak/chrontext#hasTimestamp> ?t)
